File: RemindDemo/app/src/main/java/com/example/reminddemo/base/BaseActivity.java
package com.example.reminddemo.base;

import android.os.Bundle;
import android.os.PersistableBundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.reminddemo.utils.ClassUtil;

public abstract class BaseActivity<VM extends ViewModel, DB extends ViewDataBinding> extends AppCompatActivity {

    protected abstract int getLayoutId();
    protected VM viewModel;
    protected DB bindingView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        bindingView = DataBindingUtil.setContentView(this, getLayoutId());
        bindingView.setLifecycleOwner(this);

        try {
            getSupportActionBar().hide();
        }catch (Exception e) {
            e.printStackTrace();
        }

        initViewModel();
    }

    protected void initViewModel(){
        Class<VM> viewModelClass = ClassUtil.getViewModel(this);
        if (viewModelClass != null) {
            viewModel = new ViewModelProvider(this).get(viewModelClass);
        }
    }

    public VM getViewModel() {
        return viewModel;
    }

}

File: KotlinDemo/app/src/main/java/com/example/kotlindemo/base/BaseActivity.kt
package com.example.kotlindemo.base

import android.os.Bundle
import android.os.PersistableBundle
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import com.example.kotlindemo.modules.login.impl.LoginPresenterImpl

abstract class BaseActivity<P> : AppCompatActivity() where P : IBasePresenter{

    lateinit var presenter: P

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = createP()

    }

    fun hideActionBar() {
        //任何Java代码东西，必须使用 ? 允许为null，来接受，防止空指针异常
        val actionBar: ActionBar? = supportActionBar
        actionBar?.hide()
    }

    fun showActionBar() {
        supportActionBar?.show()
    }

    override fun onDestroy() {
        super.onDestroy()
        recycle()
    }
    abstract fun createP(): P
    abstract fun recycle()
}